src/Analyzer/Passes/FunctionToSubcolumnsPass.cpp

```diff
@@ -16,218 +17,419 @@
+
+    if (!storage->supportsOptimizationToSubcolumns() || storage->isVirtualColumn(column.name, storage_snapshot->metadata))
+        return {};
+
+    auto column_in_table = storage_snapshot->tryGetColumn(GetColumnsOptions::All, column.name);
+    if (!column_in_table || !column_in_table->type->equals(*column.type))
+        return {};
+
+    return std::make_tuple(function_node, first_argument_column_node, table_node);
+}
+
+/// First pass collects info about identifiers to determine which identifiers are allowed to optimize.
+class FunctionToSubcolumnsVisitorFirstPass : public InDepthQueryTreeVisitorWithContext<FunctionToSubcolumnsVisitorFirstPass>
 {
 public:
-    using Base = InDepthQueryTreeVisitorWithContext<FunctionToSubcolumnsVisitor>;
+    using Base = InDepthQueryTreeVisitorWithContext<FunctionToSubcolumnsVisitorFirstPass>;
     using Base::Base;
 
-    void enterImpl(QueryTreeNodePtr & node) const
+    struct Data
+    {
+        bool has_final = false;
+        std::unordered_set<Identifier> all_key_columns;
+        std::unordered_map<Identifier, UInt64> indentifiers_count;
+        std::unordered_map<Identifier, UInt64> optimized_identifiers_count;
+    };
+
+    Data getData() const { return data; }
+
+    void enterImpl(const QueryTreeNodePtr & node)
     {
         if (!getSettings().optimize_functions_to_subcolumns)
             return;
 
-        auto * function_node = node->as<FunctionNode>();
-        if (!function_node)
+        if (data.has_final)
             return;
 
-        auto & function_arguments_nodes = function_node->getArguments().getNodes();
-        size_t function_arguments_nodes_size = function_arguments_nodes.size();
-
-        if (function_arguments_nodes.empty() || function_arguments_nodes_size > 2)
+        if (auto * table_node = node->as<TableNode>())
+        {
+            enterImpl(*table_node);
             return;
+        }
 
-        auto * first_argument_column_node = function_arguments_nodes.front()->as<ColumnNode>();
-
-        if (!first_argument_column_node)
+        if (auto * column_node = node->as<ColumnNode>())
+        {
+            enterImpl(*column_node);
             return;
+        }
 
-        if (first_argument_column_node->getColumnName() == "__grouping_set")
+        auto [function_node, first_argument_node, table_node] = getTypedNodesForOptimization(node);
+        if (function_node && first_argument_node && table_node)
+        {
+            enterImpl(*function_node, *first_argument_node, *table_node);
             return;
+        }
+    }
 
-        auto column_source = first_argument_column_node->getColumnSource();
-        auto * table_node = column_source->as<TableNode>();
+private:
+    Data data;
+    NameSet processed_tables;
 
-        if (!table_node)
+    void enterImpl(const TableNode & table_node)
+    {
+        if (table_node.hasTableExpressionModifiers() && table_node.getTableExpressionModifiers()->hasFinal())
+        {
+            data.has_final = true;
```

Comment:
Maybe it's better to take a look at column sources? I suspect you won't optimize the following query:

```sql
SELECT length(x) FROM t1
WHERE length(x) in (SELECT y FROM t2 FINAL)
```

Reply:
Fixed